### DISCUSSION TITLE
Wait for TCP message in agent reflex

### DISCUSSION THREAD

**Discussion by ivan-majic** [gama-platform/gama#76] [Q&A] [answered]
Hi,

I have a simulation where tourists are visiting different points of interest and I have a trained ML classifier in Python that predicts the next destination based on tourist history.

My people agents have a reflex where they need to select the next POI to visit, and this is where I would like to:
1. send agent history to a running Python TCP server (I got this to work)
2. wait for a message from the Python TCP server that contains the predicted next destination
3. make the agent go to the predicted destination

Right now I have this code in my agent reflex and it just stalls infinitely while waiting for a message:

```
reflex visit when: objective = "visiting" and (current_date.hour + current_date.minute/60) >= start_visiting {
    do send contents: '{"people_id": "'+self+'", "ktyp":"'+ktyp+'", "date":"'+current_date+'","visited_pois":'+visited_places_names+'}';

    the_target <- nil;

    loop while: the_target=nil { 
        do fetch_message_from_network;//forces gama to get all the new messages since the begining of the cycle
        if has_more_message() {
            message mess <- fetch_message();
            the_target <- mess.contents;
            write the_target;
        }
    }
    ...
```

If anyone here has experience with TCP communications in GAMA, I would greatly appreciate some insight into what am I doing wrong.

Thanks!

**Comment by lesquoyb** [comment, marked as answer]
Hello, 

I don't see anything wrong with your code, the two points that I think can cause problem would be the initialization of the connection or the data/way you send data to gama in your python server. 
Can you try again with a different connection protocol in gama (I think you need to use "tcp_server" protocol to receive AND send data) and/or by adding a "\n" and make sure you flush your message in python ?

We have some example codes available [in the documentation](https://gama-platform.org/wiki/next/UsingNetwork) in case you missed it.

If it still doesn't work you can send an example of your gaml and python server here and I'll have a closer look at it

**Reply by ivan-majic** [reply]
Hi @lesquoyb , thank you for your insightful answer.
Adding the "\n" at the end of the message in Python solved the issue and now the code works as expected.
